core/hotspot/traffic_shaping.go

```diff
@@ -155,6 +159,73 @@ func (c *baseTrafficShapingController) BoundParamIndex() int {
 	return c.paramIndex
 }
 
+// ExtractArgs matches the arg from ctx based on TrafficShapingController
+// return nil if match failed.
+func (c *baseTrafficShapingController) ExtractArgs(ctx *base.EntryContext) []interface{} {
+	if c == nil {
+		return nil
+	}
+	args := make([]interface{}, 0)
+
+	indexArg := c.extractArgs(ctx)
+	if indexArg != nil {
+		args = append(args, indexArg)
+	}
+	indexArg = c.extractAttachmentArgs(ctx)
+	if indexArg != nil {
+		args = append(args, indexArg)
+	}
+	return args
+}
+func (c *baseTrafficShapingController) extractArgs(ctx *base.EntryContext) interface{} {
+	args := ctx.Input.Args
+	idx := c.BoundParamIndex()
+	if idx < 0 {
+		idx = len(args) + idx
+	}
+	if idx < 0 {
+		if logging.DebugEnabled() {
+			logging.Debug("[extractArgs] The param index of hotspot traffic shaping controller is invalid",
+				"args", args, "paramIndex", c.BoundParamIndex())
+		}
+		return nil
+	}
+	if idx >= len(args) {
+		if logging.DebugEnabled() {
+			logging.Debug("[extractArgs] The argument in index doesn't exist",
+				"args", args, "paramIndex", c.BoundParamIndex())
+		}
+		return nil
+	}
+	return args[idx]
+}
+func (c *baseTrafficShapingController) extractAttachmentArgs(ctx *base.EntryContext) interface{} {
+	attachments := ctx.Input.Attachments
+
+	if attachments == nil {
+		if logging.DebugEnabled() {
+			logging.Debug("[extractAttachment] The attachments of ctx is nil",
+				"args", attachments, "paramIndex", c.paramKey)
+		}
+	}
+	if c.paramKey == "" {
+		if logging.DebugEnabled() {
+			logging.Debug("[extractAttachment] The param key is nil",
+				"args", attachments, "paramIndex", c.paramKey)
```

Comment:
The key of the log should be paramKey

Reply:
![image](https://user-images.githubusercontent.com/13758306/102847381-2fcbfb80-444d-11eb-991f-c2909bba4a95.png)